### ISSUE TITLE
Build fail with OpenCV

### ISSUE THREAD

**Issue by mzhang001** [apache/incubator-mxnet#2666]
# When I build mxnet with OpenCV, it fails.

gcc/4.9.2
cuda/7.5
OpenCV/2.4.13
OpenBLAS/0.2.15

echo $PKG_CONFIG_PATH
/usr/local/Cluster-Apps/opencv/2.4.13/lib/pkgconfig

echo $LD_LIBRARY_PATH
# /usr/local/Cluster-Apps/gcc/4.9.2/lib64:/usr/local/Cluster-Apps/gcc/4.9.2/lib:/usr/local/Cluster-Apps/OpenBLAS/0.2.15/lib:/usr/local/Cluster-Apps/cuda/7.5/lib64:/usr/local/Cluster-Apps/cuda/7.5/lib:/usr/local/Cluster-Apps/opencv/2.4.13/lib:/usr/local/Cluster-Apps/python/2.7.10/wx/2.9.5.0/lib:/usr/local/Cluster-Apps/python/2.7.10/lib:/usr/local/Cluster-Apps/vtk/5.6.1/lib/vtk-5.6:/usr/local/Cluster-Apps/intel/mkl/10.3.10.319/composer_xe_2011_sp1.10.319/mkl/lib/intel64:/usr/local/Cluster-Apps/intel/fce/12.1.10.319/lib/intel64:/usr/local/Cluster-Apps/intel/cce/12.1.10.319/ipp/lib/intel64:/usr/local/Cluster-Apps/intel/cce/12.1.10.319/tbb/lib/intel64/cc4.1.0_libc2.4_kernel2.6.16.21:/usr/local/Cluster-Apps/intel/cce/12.1.10.319/lib/intel64:/usr/local/Cluster-Apps/global/lib:/usr/local/Cluster-Apps/intel/impi/4.1.3.045/lib64:/usr/local/Cluster-Apps/vgl/2.3.1/64/lib:/usr/local/Cluster-Apps/slurm/lib

g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I/home/mz342/workspace/mxnet/mshadow/ -I/home/mz342/workspace/mxnet/dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMSDHADOW_USE_PASCAL=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -fopenmp -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/image_aug_default.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/broadcast_reduce_op.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/cudnn_convolution.o build/src/operator/custom.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_broadcast_op.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/elementwise_unary_op.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/l2_normalization.o build/src/operator/leaky_relu.o build/src/operator/loss_binary_op.o build/src/operator/lrn.o build/src/operator/make_loss.o build/src/operator/matrix_op.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/operator_util.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/roi_pooling.o build/src/operator/sample_op.o build/src/operator/slice_channel.o build/src/operator/smooth_l1_unary.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/spatial_transformer.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/graph_memory_allocator.o build/src/symbol/static_graph.o build/src/symbol/symbol.o /home/mz342/workspace/mxnet/dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/broadcast_reduce_op_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_broadcast_op_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/elementwise_unary_op_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/l2_normalization_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/loss_binary_op_gpu.o build/src/operator/lrn_gpu.o build/src/operator/make_loss_gpu.o build/src/operator/matrix_op_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/roi_pooling_gpu.o build/src/operator/sample_op_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/smooth_l1_unary_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/spatial_transformer_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -lopenblas -fopenmp -lrt `pkg-config --libs opencv` -lcudnn -lopencv_core -lopencv_highgui -lopencv_features2d -lopencv_flann -lopencv_imgproc -lcuda
/tmp/cc3XsZJN.o: In function `main':
/home/mz342/workspace/mxnet/tools/im2rec.cc:236: undefined reference to`cv::imencode(std::string const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to `vtable for std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find(char const_, unsigned long, unsigned long) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_highgui.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::assign(char const_)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_M_create(unsigned long&, unsigned long)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_erase(unsigned long, unsigned long)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_flann.so: undefined reference to`std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream(std::_Ios_Openmode)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace_aux(unsigned long, unsigned long, unsigned long, char)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to`std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_stringstream()@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::rfind(char, unsigned long) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const_, unsigned long)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::reserve(unsigned long)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::resize(unsigned long, char)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to `vtable for std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find(char, unsigned long) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find_last_of(char const_, unsigned long, unsigned long) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace(unsigned long, unsigned long, char const*, unsigned long)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_flann.so: undefined reference to`std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_ostringstream()@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::_M_sync(char*, unsigned long, unsigned long)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::swap(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_ocl.so: undefined reference to `std::basic_istream<char, std::char_traits<char> >& std::getline<char, std::char_traits<char>, std::allocator<char> >(std::basic_istream<char, std::char_traits<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, char)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_flann.so: undefined reference to`std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::str() const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_core.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find_first_not_of(char const*, unsigned long, unsigned long) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to`VTT for std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_highgui.so: undefined reference to`std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream(std::_Ios_Openmode)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_ocl.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::append(char const*)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_contrib.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::compare(char const_) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_videostab.so: undefined reference to `std::runtime_error::runtime_error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_highgui.so: undefined reference to`vtable for std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_highgui.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct(unsigned long, char)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_flann.so: undefined reference to`std::runtime_error::runtime_error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_highgui.so: undefined reference to `VTT for std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_flann.so: undefined reference to`std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::compare(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const@GLIBCXX_3.4.21'
/usr/local/Cluster-Apps/opencv/2.4.13/lib/libopencv_calib3d.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create(unsigned long&, unsigned long)@GLIBCXX_3.4.21'
collect2: error: ld returned 1 exit status
make: *_\* [bin/im2rec] Error 1

Thanks a lot

**Comment by piiswrong** [comment]
Looks like your opencv is not correctly compiled & installed. Are you able to like it to other libraries?

**Comment by mzhang001** [comment]
like it?

**Comment by mzhang001** [comment]
you mean whether I can link OpenCV to other libraries?

**Comment by ranjanan** [comment]
Try following the instructions [here](http://mxnet.io/get_started/setup.html?highlight=opencv) to build OpenCV from source. That should help.

**Comment by pbertoni** [comment]
Maybe [this](https://github.com/opencv/opencv/issues/9446) build-environment related issue could help.

**Comment by szha** [comment]
All mxnet pre-built pip packages are built with opencv support. See instructions here. https://mxnet.incubator.apache.org/get_started/install.html

Feel free to ping me to reopen if this is still an active issue.
